README.md

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

# Dockerized OpenTOSCA Environment

> Docker Compose file for running the entire OpenTOSCA stack.

The fastest way to get started is using [Docker Compose](https://docs.docker.com/compose/):

    docker-compose up

Wait a few seconds, then open the [OpenTOSCA user interface](http://localhost:8088).

| OpenTOSCA Component | URL | GitHub | Docker Hub |
|:------------------- |:--- |:------ |:---------- |
| OpenTOSCA UI | <http://localhost><br><http://localhost:8088> | [Link](https://github.com/OpenTOSCA/ui) | [Link](https://hub.docker.com/r/opentosca/ui) |
| OpenTOSCA Modelling (Eclipse Winery) | <http://localhost/winery><br><http://localhost:8080> | [Link](https://github.com/OpenTOSCA/winery) | [Link](https://hub.docker.com/r/opentosca/winery) |
| OpenTOSCA Container API | <http://localhost:1337> | [Link](https://github.com/OpenTOSCA/container) | [Link](https://hub.docker.com/r/opentosca/container) |
| OpenTOSCA Container Repository | <http://localhost:8081> | [Link](https://github.com/OpenTOSCA/winery) | [Link](https://hub.docker.com/r/opentosca/winery) |
| Plan Engine (Apache ODE) | <http://localhost/ode><br><http://localhost:9763/ode> | [Link](https://github.com/OpenTOSCA/ode) | [Link](https://hub.docker.com/r/opentosca/ode) |
| IA Engine (Apache Tomcat) | <http://localhost/manager><br><http://localhost:8090/manager><br>(user: `admin`, password: `admin`) | [Link](https://github.com/OpenTOSCA/engine-ia) | [Link](https://hub.docker.com/r/opentosca/engine-ia) |

**Make sure following ports in your environment are free in order to start OpenTOSCA properly:**

* `1337`
* `8080-8088`
* `8090`
* `9763`
* `1883`
* `9000` (optional)

> It is recommended that your host or virtual machine has at least 4GB of memory.

**NOTE:** Please check the [Docker Daemon Settings](#docker-daemon-settings)

**NOTE:** In an optimal setup `localhost` should be replaced by a publicly available Fully-Qualified Domain Name or ip address.

---

## How To

Simple How-To section to cover different kinds of use cases.

> **Info:** We use the override feature of Docker Compose to provide different configurations for certain use cases.
> [More information](https://docs.docker.com/compose/extends).
>
> Basic override file with common configuration settings for our environment: [`_docker-compose.override.yml`](_docker-compose.override.yml).
> Simply, make a copy and modify it to your needs:
> ```shell
> cp _docker-compose.override.yml docker-compose.override.yml
> ```
> Settings from `docker-compose.override.yml` are applied automatically when using `docker-compose up`.

### How to use an existing Winery repository

You can map an existing Winery repository (on your host) as a volume into the `winery` container.

* Create a `docker-compose.override.yml` file (or copy it from `_docker-compose.override.yml`)
* Use the following configuration in order to map a repository:

  ```yaml
    winery:
      volumes:
        - <path on host system>:/var/opentosca/repository
  ```

* Replace `<path on host system>` with the path to a valid Winery repository (can be empty)
* Start the environment as usual: `docker-compose up -d`

### How to run the environment with WSO2 BPS engine

Start the environment with the `docker-compose.bps.yml` override:

```shell
docker-compose -f docker-compose.yml -f docker-compose.bps.yml up -d
```

### How to run the environment in production (Linux only)

* Follow the installation instruction for Docker and Docker Compose (Linux):
  * [Docker CE](https://docs.docker.com/install/linux/docker-ce/ubuntu)
  * [Docker Compose](https://docs.docker.com/compose/install)
* Create the following directories (required to map volumes properly):

  ```shell
  mkdir -p /var/opentosca/container/data
  mkdir -p /var/opentosca/container/repository
  mkdir -p /var/opentosca/winery/repository
  mkdir -p /var/opentosca/portainer
  ```

* Open the `.env` file and set the `PUBLIC_HOSTNAME` variable to your host's public IP address or hostname
* Start the environment

  ```shell
  docker-compose -f docker-compose.yml -f docker-compose.prod.yml up -d
  ```

* This setup also starts the lightweight management UI **Portainer** on port `9000`
* For special scenarios, you may have to specify an override file:

  ```shell
  docker-compose.exe -f docker-compose.yml -f docker-compose.prod.yml -f docker-compose.override.yml up -d
  ```

### How to debug components running inside the environment

* Create a `docker-compose.override.yml` file (or copy it from `_docker-compose.override.yml`)
* Make sure you enable and set the certain Java debug options (e.g., `-Xdebug`)
* Start the environment as usual: `docker-compose up -d`

### How to clone a private TOSCA definitions repository to be used with Winery

* Start the environment as usual: `docker-compose up -d`

* Log in to the Winery container and clone the respective Git repository manually (we use our internal `tosca-definitions-internal` repo as an example here):

  ```shell
  docker-compose exec winery bash
  rm -rf /var/opentosca/repository
  git clone https://github.com/OpenTOSCA/tosca-definitions-internal /var/opentosca/repository
  exit
  ```

* Open Winery at <http://localhost:8080>
  > **Note:** Changes to the repository has to be pushed manually with `git push`.

### How to extend Winery's JVM Heap Size

You can adjust Winery's JVM heap size by setting a respective environment variable for the `winery` container.

* Create a `docker-compose.override.yml` file (or copy it from `_docker-compose.override.yml`)
* Use the following configuration in order to map a repository:

  ```yaml
    winery:
      environment:
        WINERY_HEAP_MAX: 8192m
  ```

* **Note:** Make sure the hosting VM has enough capacity
* Start the environment as usual: `docker-compose up -d`

### Tips and Tricks

#### General

```bash
# Pull the latest images
docker-compose pull

# Validate and view the resulting configuration
docker-compose [-f <file> ...] config

# Start services in background
docker-compose up -d

# Shutdown services and remove container
docker-compose down -v

# Display useful logs
docker-compose logs -f [--tail=1 <SERVICE_NAME>...]
docker-compose logs -f container
docker-compose logs -f engine-ia engine-plan
```

#### Docker Daemon Settings

For a good user experience set up your Docker environment accordingly:

![Docker MAC Seetings](docs/docker_daemon.png)

---

## Haftungsausschluss

Dies ist ein Forschungsprototyp.
Die Haftung für entgangenen Gewinn, Produktionsausfall, Betriebsunterbrechung, entgangene Nutzungen, Verlust von Daten und Informationen, Finanzierungsaufwendungen sowie sonstige Vermögens- und Folgeschäden ist, außer in Fällen von grober Fahrlässigkeit, Vorsatz und Personenschäden ausgeschlossen.

## Disclaimer of Warranty

Unless required by applicable law or agreed to in writing, Licensor provides the Work (and each Contributor provides its Contributions) on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied, including, without limitation, any warranties or conditions of TITLE, NON-INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A PARTICULAR PURPOSE.
You are solely responsible for determining the appropriateness of using or redistributing the Work and assume any risks associated with Your exercise of permissions under this License.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Added

- Added proxy to reach frontends using `<hostname>:80/(ui|winery|ode|manager)`

## [2.1.0] (Silvaner) - 2018-08-16

> Initial submission by @jojow: https://github.com/jojow/opentosca-dockerfiles

### Added

- Added override configuration to use WSO2 BPS.
- Added production-ready configuration for Linux-based operating systems.
- Added basic override configuration. Can be used with `cp _docker-compose.override.yml docker-compose.override.yml`.
- Added a Changelog file based on the format of [Keep a Changelog](http://keepachangelog.com/).
- Added build arg `GIT_REPO_URL` to configure the used Git repository.
- The containers (`winery`, `web`) are now able to be run individually, without the requirement to have a linked Tomcat image in place.
- Eclipse Winery™ is used as Container Repository as well as Modelling Tool in separate containers.
- Docker Compose configuration uses a `bridge` network in order to use service names as FQDNs.
- Building `opentosca-ui` container with new Angular-based OpenTOSCA UI.
- Building container for BMPN4TOSCA Modeler (`winery-bpmn4tosca').
- Docker Compose: Added Docker-in-Docker container (`jpetazzo/dind`) to run Docker-based topologies with OpenTOSCA.

### Changed

- Moved the `Dockerfiles` to dedicated repositories; building and publishing images to Docker Hub is now part of those repositories.
- Changed default plan engine to Apache ODE.
- Build Tomcat container from official image.
- Build BPS container from official OpenJDK image.
- Build `opentosca-container` container from official Maven image.
- Build Winery container from official Maven image.
- Docker Compose: Variable `GIT_BRANCH` can be used to build containers from a specific Git branch.

### Removed

- Removed Vinothek container since it will be replaced with the new OpenTOSCA UI (`opentosca/ui`).
- Removed `ROOT.war`; it will be replaced with a new landing page in future.

[Unreleased]: https://github.com/OpenTOSCA/opentosca-docker/compare/v2.1.0...HEAD
[2.1.0]: https://github.com/OpenTOSCA/opentosca-docker/releases/tag/v2.1.0


runner/post/post.js
'use strict';

const fs = require('fs');
const path = require('path');
const async = require('async');
const _ = require('lodash');
const got = require('got');
const FormData = require('form-data');

const ADMIN_UI_BASE = 'http://localhost:8080';
//const CSARS_DIR = '/csars';
//const ENDPOINTS_JSON = '/endpoints/endpoints.json';
const UNDEPLOY_TIMEOUT = 2 * 60 * 1000;

let csars = [];



async.series([
  function(done) {
    got(ADMIN_UI_BASE + '/admin/getCSARList.action', {
      json: true
    }).then(response => {
      csars = response.body.csars;

      console.log('CSARs to be undeployed', csars);

      done();
    }).catch(err => {
      done(err);
    });
  },
  //TODO terminate instances
  // GET http://localhost:8080/admin/getCSARInstanceList.action?selectedCSAR=test-wrapped.csar
  // RESPONSE JSON: {"availableInstances":[],"selectedCSAR":"test-wrapped.csar"}
  function(done) {
    async.eachSeries(csars, function(csar, done) {
      console.log('undeploying CSAR', csar);

      const form = new FormData();

      form.append('selectedCSAR', csar + '.csar');

      got.post(ADMIN_UI_BASE + '/admin/deleteCSAR.action', {
        timeout: UNDEPLOY_TIMEOUT,
        headers: form.getHeaders(),
        body: form
      }).then(response => {
        console.log('CSAR undeployed', csar);

        done();
      }).catch(err => {
        done(err);
      });
    }, done);
  }
], (err) => {
  if (err) throw err;
});


runner/pre/pre.js
'use strict';

const fs = require('fs');
const path = require('path');
const glob = require('glob');
const async = require('async');
const _ = require('lodash');
const got = require('got');
const FormData = require('form-data');

const ENGINE_API_BASE = 'http://localhost:1337';
const ADMIN_UI_BASE = 'http://localhost:8080';
const BPS_UI_BASE = 'https://localhost:9443';
const CSARS_DIR = '/csars';
const OPENTOSCA_ENDPOINTS_JSON = '/endpoints/opentosca.json';
const ENDPOINTS_JSON = '/endpoints/endpoints.json';
const RETRY_DELAY = 15 * 1000;
const DEPLOY_TIMEOUT = 45 * 60 * 1000;

let csars = [];
let endpoints = {};



async.series([
  function(done) {
    glob(CSARS_DIR + '/*', {
      nodir: true,
      dot: false
    }, (err, files) => {
      if (err) return done(err);

      csars = files;

      console.log('CSARs to be deployed', csars);

      done();
    });
  },
  function(done) {
    console.log('waiting for OpenTOSCA APIs to be available');

    setTimeout(done, RETRY_DELAY);
  },
  function(done) {
    let retries = 0;
    const maxRetries = 12;

    let uiUp = false;
    let apiUp = false;
    let bpsUp = false;

    async.whilst(function() {
      return (!uiUp || !apiUp || !bpsUp) && retries < maxRetries;
    }, function(done) {
      async.series([
        function(done) {
          setTimeout(done, RETRY_DELAY);
        },
        function(done) {
          if (uiUp) return done();

          got(ADMIN_UI_BASE + '/admin/getCSARList.action').then(response => {
            retries++;

            if (response.statusCode !== 200) return done();

            uiUp = true;

            console.log('OpenTOSCA admin UI available');

            done();
          }).catch(err => {
            retries++;

            done();
          });
        },
        function(done) {
          if (apiUp || !uiUp) return done();

          got(ENGINE_API_BASE + '/containerapi/CSARs').then(response => {
            if (response.statusCode !== 200) return done();

            apiUp = true;

            console.log('OpenTOSCA container API available');

            done();
          }).catch(err => {
            done();
          });
        },
        function(done) {
          if (bpsUp || !uiUp || !apiUp) return done();

          got(BPS_UI_BASE + '/carbon/admin/login.jsp', {
            rejectUnauthorized: false
          }).then(response => {
            if (response.statusCode !== 200) return done();

            bpsUp = true;

            console.log('OpenTOSCA BPS available');

            done();
          }).catch(err => {
            done();
          });
        }
      ], done);
    }, (err) => {
      if (err) done(err);
      else if (!uiUp) done(new Error('OpenTOSCA admin UI not available'));
      else if (!apiUp) done(new Error('OpenTOSCA container API not available'));
      else if (!bpsUp) done(new Error('OpenTOSCA BPS not available'));
      else done();
    });
  },
  function(done) {
    console.log('additional delay, required to ensure OpenTOSCA is ready for CSAR deployment');

    setTimeout(done, RETRY_DELAY);
  },
  function(done) {
    async.eachSeries(csars, function(csar, done) {
      console.log('deploying CSAR', csar);

      const form = new FormData();

      // path.basename(csar, '.csar')
      form.append('file', fs.createReadStream(csar));

      const timer = setInterval(() => {
        console.log('still deploying CSAR', csar);
      }, 5 * 60 * 1000);

      got.post(ADMIN_UI_BASE + '/admin/uploadCSAR.action', {
        timeout: DEPLOY_TIMEOUT,
        headers: form.getHeaders(),
        body: form
      }).then(response => {
        clearInterval(timer);

        done();
      }).catch(err => {
        clearInterval(timer);

        if (err.response.statusCode !== 302) return done(err);

        console.log('CSAR deployed', csar);

        done();
      });
    }, done);
  },
  function(done) {
    fs.access(OPENTOSCA_ENDPOINTS_JSON, fs.R_OK, (err) => {
      if (err) {
        console.log('cannot find opentosca-endpoints', OPENTOSCA_ENDPOINTS_JSON);

        return done();
      }

      console.log('opentosca-endpoints file found', OPENTOSCA_ENDPOINTS_JSON);

      fs.readFile(OPENTOSCA_ENDPOINTS_JSON, 'utf8', (err, content) => {
        if (err) return done(err);

        _.forEach(_.map(content.trim().split('\n'), JSON.parse), (entry) => {
          if (!entry.deployed || entry.undeployed) {
            delete endpoints[entry.endpoint];

            return;
          }

          _.forEach(entry, (val, key) => {
            if (val === 'null') entry[key] = null;
          });

          endpoints[entry.endpoint] = entry;
        });

        console.log('endpoints consolidated', endpoints);

        done();
      });
    });
  },
  function(done) {
    fs.writeFile(ENDPOINTS_JSON, JSON.stringify(endpoints, null, 2), done);
  }
], (err) => {
  if (err) throw err;
});



// POST http://localhost:8080/admin/uploadCSAR.action
// Content-Type: multipart/form-data, app octet-stream

// GET http://localhost:8080/admin/getCSARList.action
// {"csars":["test-wrapped"],"selectedCSAR":null}

// GET http://docker:8080/admin/getCSARInstanceList.action?selectedCSAR=test-wrapped.csar
// RESPONSE JSON: {"availableInstances":[],"selectedCSAR":"test-wrapped.csar"}

// POST http://localhost:8080/admin/deleteCSAR.action
// HEADER: Content-Type: application/x-www-form-urlencoded
// BODY: form: selectedCSAR=test-wrapped.csar

// GET http://localhost:1337/containerapi/CSARs/moodle2.csar/Content/SELFSERVICE-Metadata/plan.input.default.xml
// ERROR: cannot get file ... try accept(APPLICATION_OCTET_STREAM)
// ... read from Container API: CSARs/SELFSERVICE-Metadata/data.xml -> reference to planInputFile

// GET http://localhost:8080/vinothek/ApplicationInstantiation?container=localhost&applicationId=http://localhost:1337/containerapi/CSARs/moodle2.csar/Content/SELFSERVICE-Metadata/&optionId=1&xml=%3Csoapenv%3AEnvelope+xmlns%3Asoapenv%3D%22http%3A%2F%2Fschemas.xmlsoap.org%2Fsoap%2Fenvelope%2F%22+xmlns%3Aorg%3D%22http%3A%2F%2F%2Fwww.opentosca.org%2Fexamples%2FMoodle%2FBuildPlan%22%3E%0A+++%3Csoapenv%3AHeader%2F%3E%0A+++%3Csoapenv%3ABody%3E%0A++++++%3Corg%3AMoodleBuildPlanRequest%3E%0A+++++++++%3Corg%3Aregion%3EAWS-REGION-NAME+(e.g.%2C+%22ec2.eu-west-1.amazonaws.com%22)%3C%2Forg%3Aregion%3E%0A+++++++++%3Corg%3AsecurityGroup%3EAWS-SECURITY-GROUP-NAME+(All+TCP+ports+open)%3C%2Forg%3AsecurityGroup%3E%0A+++++++++%3Corg%3AkeyPairName%3EAWS-KEY-PAIR-NAME%3C%2Forg%3AkeyPairName%3E%0A+++++++++%3Corg%3AsshKey%3E-----BEGIN+RSA+PRIVATE+KEY-----%0A...%0AAWS-SSH-KEY%0A(Ensure+that+all+line+breaks+are+conserved+and+no+white+spaces+at+the+beginning+of+the+lines+are+added!)%0A...%0Aabcdefghijklmnopqrstuvwxyz0123456789%3D%3D%0A-----END+RSA+PRIVATE+KEY-----%3C%2Forg%3AsshKey%3E%0A+++++++++%3Corg%3Aami%3EAWS-AMI-ID+(e.g.%2C+%22ami-ce7b6fba%22)%3C%2Forg%3Aami%3E%0A+++++++++%3Corg%3AinstanceType%3Et1.micro%3C%2Forg%3AinstanceType%3E%0A+++++++++%3Corg%3AaccessKey%3EAWS-ACCESS-KEY%3C%2Forg%3AaccessKey%3E%0A+++++++++%3Corg%3AsecretKey%3EAWS-SECRET-KEY%3C%2Forg%3AsecretKey%3E%0A%09%09+%3C!--+Example+for+%25CSAR-NAME%25+%3D+Moodle.csar+--%3E%0A%09%09+%3Corg%3AcsarName%3E%25CSAR-NAME%25%3C%2Forg%3AcsarName%3E%0A%09%09+%3C!--+Example+for+%25CONTAINER-API%25+%3D+http%3A%2F%2F%3CCONTAINER-HOST%3E%3A1337%2Fcontainerapi+--%3E%0A%09%09+%3Corg%3AcontainerApi%3E%25CONTAINER-API%25%3C%2Forg%3AcontainerApi%3E%0A%09%09+%3Corg%3AcallbackUrl%3E%25CALLBACK-URL%25%3C%2Forg%3AcallbackUrl%3E%0A%09%09+%3Corg%3ACorrelationID%3E%25CORRELATION-ID%25%3C%2Forg%3ACorrelationID%3E%0A++++++%3C%2Forg%3AMoodleBuildPlanRequest%3E%0A+++%3C%2Fsoapenv%3ABody%3E%0A%3C%2Fsoapenv%3AEnvelope%3E
/* QUERY STRING PARAMS URL-DECODED:
container=localhost
applicationId=http://localhost:1337/containerapi/CSARs/moodle2.csar/Content/SELFSERVICE-Metadata/
optionId=1
xml=<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:org="http:///www.opentosca.org/examples/Moodle/BuildPlan">
   <soapenv:Header/>
   <soapenv:Body>
      <org:MoodleBuildPlanRequest>
         <org:region>AWS-REGION-NAME (e.g., "ec2.eu-west-1.amazonaws.com")</org:region>
         <org:securityGroup>AWS-SECURITY-GROUP-NAME (All TCP ports open)</org:securityGroup>
         <org:keyPairName>AWS-KEY-PAIR-NAME</org:keyPairName>
         <org:sshKey>-----BEGIN RSA PRIVATE KEY-----
...
AWS-SSH-KEY
(Ensure that all line breaks are conserved and no white spaces at the beginning of the lines are added!)
...
abcdefghijklmnopqrstuvwxyz0123456789==
-----END RSA PRIVATE KEY-----</org:sshKey>
         <org:ami>AWS-AMI-ID (e.g., "ami-ce7b6fba")</org:ami>
         <org:instanceType>t1.micro</org:instanceType>
         <org:accessKey>AWS-ACCESS-KEY</org:accessKey>
         <org:secretKey>AWS-SECRET-KEY</org:secretKey>
		 <!-- Example for %CSAR-NAME% = Moodle.csar -->
		 <org:csarName>%CSAR-NAME%</org:csarName>
		 <!-- Example for %CONTAINER-API% = http://<CONTAINER-HOST>:1337/containerapi -->
		 <org:containerApi>%CONTAINER-API%</org:containerApi>
		 <org:callbackUrl>%CALLBACK-URL%</org:callbackUrl>
		 <org:CorrelationID>%CORRELATION-ID%</org:CorrelationID>
      </org:MoodleBuildPlanRequest>
   </soapenv:Body>
</soapenv:Envelope> */
// RESPONSE STRING: http://localhost:8080/vinothek/CallbackStatus?callbackId=1471439852916

// GET http://localhost:8080/vinothek/CallbackStatus?callbackId=1471439852916
// RESPONSE JSON: {"NO-CALLBACK-RECEIVED-YET":true}
